Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+    }
+
+    thread_data = NULL;
+    for (int i = 0; i < MAX_SEMA_DEPTH; i++) {
+        _PySemaphore_Destroy(&td->semas[i]);
+    }
+
+    PyMem_RawFree(td);
+}
+
+static void
+enqueue(Bucket *bucket, const void *address, struct wait_entry *wait)
+{
+    if (!bucket->root.next) {
+        // initialize bucket
+        llist_init(&bucket->root);
+    }
+    llist_insert_tail(&bucket->root, &wait->node);
+    ++bucket->num_waiters;
+}
+
+static struct wait_entry *
+dequeue(Bucket *bucket, const void *address)
+{
+    struct llist_node *root = &bucket->root;
+    if (!root->next) {
+        // bucket was not yet initialized
+        return NULL;
+    }
+
+    // find the first waiter that is waiting on `address`
+    struct llist_node *node;
+    llist_for_each(node, root) {
+        struct wait_entry *wait = llist_data(node, struct wait_entry, node);
+        if (wait->addr == (uintptr_t)address) {
+            llist_remove(node);
+            --bucket->num_waiters;
+            return wait;
+        }
+    }
+    return NULL;
+}
+
+static void
+dequeue_all(Bucket *bucket, const void *address, struct llist_node *dst)
+{
+    struct llist_node *root = &bucket->root;
+    if (!root->next) {
+        // bucket was not yet initialized
+        return;
+    }
+
+    // remove and append all matching waiters to dst
+    struct llist_node *node;
+    llist_for_each_safe(node, root) {
+        struct wait_entry *wait = llist_data(node, struct wait_entry, node);
+        if (wait->addr == (uintptr_t)address) {
+            llist_remove(node);
+            llist_insert_tail(dst, node);
+            --bucket->num_waiters;
+        }
+    }
+}
+
+// Checks that `*addr == *expected`
+static int
+validate_addr(const void *addr, const void *expected, size_t addr_size)
+{
+    switch (addr_size) {
+    case 1: return _Py_atomic_load_uint8(addr) == *(const uint8_t *)expected;
+    case 2: return _Py_atomic_load_uint16(addr) == *(const uint16_t *)expected;
+    case 4: return _Py_atomic_load_uint32(addr) == *(const uint32_t *)expected;
+    case 8: return _Py_atomic_load_uint64(addr) == *(const uint64_t *)expected;
+    default: Py_UNREACHABLE();
+    }
+}
+
+int
+_PyParkingLot_Park(const void *addr, const void *expected, size_t size,
+                   _PyTime_t timeout_ns, void *arg, int detach)
+{
+    struct wait_entry wait;
+    wait.unpark.arg = arg;
+    wait.addr = (uintptr_t)addr;
+
+    Bucket *bucket = &buckets[((uintptr_t)addr) % NUM_BUCKETS];
+
+    _PyRawMutex_Lock(&bucket->mutex);
+    if (!validate_addr(addr, expected, size)) {
+        _PyRawMutex_Unlock(&bucket->mutex);
+        return Py_PARK_AGAIN;
+    }
+    wait.sema = _PySemaphore_Alloc();
+    enqueue(bucket, addr, &wait);
+    _PyRawMutex_Unlock(&bucket->mutex);
+
+    int res = _PySemaphore_Wait(wait.sema, timeout_ns, detach);
+    if (res == Py_PARK_OK) {
+        goto done;
+    }
+
+    // timeout or interrupt
+    _PyRawMutex_Lock(&bucket->mutex);
+    if (wait.node.next == NULL) {
```

Comment:
As with the other cases, consider hiding the implementation detail:

```suggestion
    if (!llist_is_attached(&wait.node)) {
```

Reply:
I've made this a bit more explicit by adding a `is_unparking` field to the wait entry and use that instead of relying on the linked list nodes.

(I don't love `llist_is_attached` since an unatached node in general might have arbitrary `next` and `prev` fields, including possibly, uninitialized memory)